test/resolver_test.go

```diff
@@ -0,0 +1,194 @@
+			if cfg != nil && cfg.Context == nil {
+				cfg.Context = i.Context
+			}
+			return cfg
+		},
+	})
+	ss.r.UpdateState(state) // Blocks until config selector is applied
+
+	longdeadlineCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	shorterTimeout := 3 * time.Second
+
+	testMD := metadata.MD{"footest": []string{"bazbar"}}
+	mdOut := metadata.MD{"handler": []string{"value"}}
+
+	var onCommittedCalled bool
+
+	testCases := []struct {
+		name   string
+		md     metadata.MD
+		config *iresolver.RPCConfig
+	}{{
+		name:   "basic",
+		md:     testMD,
+		config: &iresolver.RPCConfig{},
+	}, {
+		name: "alter MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: metadata.NewOutgoingContext(ctx, mdOut),
+		},
+	}, {
+		name: "alter timeout; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: longdeadlineCtx,
+		},
+	}, {
+		name:   "nil config",
+		md:     metadata.MD{},
+		config: nil,
+	}, {
+		name: "alter timeout via method config; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			MethodConfig: serviceconfig.MethodConfig{
+				Timeout: &shorterTimeout,
+			},
+		},
+	}, {
+		name: "onCommitted callback",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			OnCommitted: func() {
+				onCommittedCalled = true
+			},
+		},
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			select {
+			case sendConfigChan <- tc.config:
+			default:
+				t.Fatalf("last config not consumed by config selector")
+			}
+
+			onCommittedCalled = false
+			ctx = metadata.NewOutgoingContext(ctx, tc.md)
+			if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+				t.Fatalf("client.EmptyCall(_, _) = _, %v; want _, nil", err)
+			}
+
+			var gotInfo *iresolver.RPCInfo
+			select {
+			case gotInfo = <-gotInfoChan:
+			default:
+				t.Fatalf("no config selector data")
+			}
+
+			var gotContext context.Context
+			select {
+			case gotContext = <-gotContextChan:
+			default:
+				t.Fatalf("no context received")
+			}
+
+			if want := "/grpc.testing.TestService/EmptyCall"; gotInfo.Method != want {
+				t.Errorf("gotInfo.Method = %q; want %q", gotInfo.Method, want)
+			}
+
+			if gotMD, _ := metadata.FromOutgoingContext(gotInfo.Context); !reflect.DeepEqual(tc.md, gotMD) {
+				t.Errorf("gotInfo.Context contains MD %v; want %v", gotMD, tc.md)
+			}
+
+			sentMD, _ := metadata.FromOutgoingContext(ctx)
+			if tc.config != nil && tc.config.Context != nil {
+				sentMD, _ = metadata.FromOutgoingContext(tc.config.Context)
+			}
+			gotMD, _ := metadata.FromIncomingContext(gotContext)
+			for k, v := range sentMD {
+				if !reflect.DeepEqual(gotMD[k], v) {
+					t.Errorf("received md = %v; want contains(%v)", gotMD, sentMD)
+				}
+			}
+
+			deadlineSent, _ := ctx.Deadline()
+			if tc.config != nil && tc.config.Context != nil {
```

Comment:
There is a lot of special casing around `config != nil` and the corresponding `context != nil`. Could you split out the tests where these are expected to be `nil` and have helper functions for shared code between the tests?